File: lib/markdown-to-note-html.ts
import { escapeHtml } from "@/lib/lesson-note-content"

function parseInlineMarkdown(text: string): string {
  let s = escapeHtml(text)
  s = s.replace(/`([^`]+)`/g, "<code>$1</code>")
  s = s.replace(/\*\*(.+?)\*\*/g, "<strong>$1</strong>")
  s = s.replace(/__(.+?)__/g, "<strong>$1</strong>")
  s = s.replace(/\*(.+?)\*/g, "<em>$1</em>")
  s = s.replace(
    /\[([^\]]+)\]\(([^)]+)\)/g,
    '<a href="$2" target="_blank" rel="noopener noreferrer">$1</a>'
  )
  return s
}

function isTableRow(line: string): boolean {
  const t = line.trim()
  return t.startsWith("|") && t.endsWith("|") && t.includes("|")
}

function isTableSeparator(line: string): boolean {
  return /^\|?[\s|:-]+\|?$/.test(line.trim()) && line.includes("-")
}

function parseTableRow(line: string): string[] {
  return line
    .trim()
    .replace(/^\|/, "")
    .replace(/\|$/, "")
    .split("|")
    .map((c) => c.trim())
}

function renderTable(lines: string[]): string {
  const rows = lines.filter((l) => l.trim() && !isTableSeparator(l))
  if (rows.length === 0) return ""

  const parsed = rows.map(parseTableRow)
  const header = parsed[0]
  const body = parsed.slice(1)

  const thead = `<thead><tr>${header.map((c) => `<th>${parseInlineMarkdown(c)}</th>`).join("")}</tr></thead>`
  const tbody = body.length
    ? `<tbody>${body
        .map(
          (row) =>
            `<tr>${row.map((c) => `<td>${parseInlineMarkdown(c)}</td>`).join("")}</tr>`
        )
        .join("")}</tbody>`
    : ""

  return `<table class="note-table">${thead}${tbody}</table>`
}

function parseBlock(block: string): string {
  const lines = block.split("\n").filter((l) => l.trim() !== "")
  if (lines.length === 0) return ""

  if (lines.every(isTableRow) || (lines.filter(isTableRow).length >= 2 && lines.some(isTableSeparator))) {
    return renderTable(lines)
  }

  const first = lines[0].trim()
  if (/^#{1,3}\s+/.test(first)) {
    const level = first.match(/^#+/)?.[0].length ?? 1
    const tag = `h${Math.min(level, 3)}`
    const text = first.replace(/^#+\s+/, "")
    return `<${tag}>${parseInlineMarkdown(text)}</${tag}>`
  }

  if (/^[-*]\s+/.test(first)) {
    const items = lines
      .filter((l) => /^[-*]\s+/.test(l.trim()))
      .map((l) => `<li>${parseInlineMarkdown(l.trim().replace(/^[-*]\s+/, ""))}</li>`)
    return items.length ? `<ul>${items.join("")}</ul>` : ""
  }

  if (/^\d+\.\s+/.test(first)) {
    const items = lines
      .filter((l) => /^\d+\.\s+/.test(l.trim()))
      .map((l) => `<li>${parseInlineMarkdown(l.trim().replace(/^\d+\.\s+/, ""))}</li>`)
    return items.length ? `<ol>${items.join("")}</ol>` : ""
  }

  if (/^>\s+/.test(first)) {
    const quote = lines.map((l) => l.trim().replace(/^>\s?/, "")).join(" ")
    return `<blockquote><p>${parseInlineMarkdown(quote)}</p></blockquote>`
  }

  if (/^```/.test(first)) {
    const code = lines.slice(1).filter((l) => !l.trim().startsWith("```")).join("\n")
    return `<pre><code>${escapeHtml(code)}</code></pre>`
  }

  const paragraph = lines.map((l) => parseInlineMarkdown(l.trim())).join("<br>")
  return `<p>${paragraph}</p>`
}

export function looksLikeMarkdown(text: string): boolean {
  const t = text.trim()
  if (!t) return false
  return (
    /^#{1,3}\s/m.test(t) ||
    /^\|.+\|$/m.test(t) ||
    /^[-*]\s+/m.test(t) ||
    /^\d+\.\s+/m.test(t) ||
    /\*\*[^*]+\*\*/.test(t) ||
    /^>\s+/m.test(t) ||
    /\[.+\]\(.+\)/.test(t)
  )
}

export function markdownToNoteHtml(markdown: string): string {
  const normalized = markdown.replace(/\r\n/g, "\n").trim()
  if (!normalized) return ""

  const blocks = normalized.split(/\n{2,}/)
  const html = blocks.map(parseBlock).filter(Boolean).join("")
  return html || `<p>${parseInlineMarkdown(normalized.replace(/\n/g, " "))}</p>`
}

export function extractTitleFromImportedContent(text: string, html: string): string | null {
  const hMatch = html.match(/<h[1-3][^>]*>([\s\S]*?)<\/h[1-3]>/i)
  if (hMatch?.[1]) {
    const plain = hMatch[1].replace(/<[^>]+>/g, "").trim()
    if (plain) return plain.slice(0, 120)
  }

  const firstLine = text.split("\n").find((l) => l.trim())?.trim() ?? ""
  if (!firstLine) return null
  const heading = firstLine.replace(/^#+\s+/, "").replace(/^[-*]\s+/, "")
  if (heading.length > 3 && heading.length <= 120) return heading
  return null
}

File: mobile/lib/utils/markdown_to_note_html.dart
import 'lesson_note_content.dart';

/// يحوّل المحتوى المخزّن (نص، Markdown، أو HTML بسيط) إلى HTML للعرض.
String resolveNoteHtmlForDisplay(String content) {
  final t = content.trim();
  if (t.isEmpty) return '<p></p>';
  final plain = isNoteHtml(t) ? htmlToEditableText(t) : t;
  if (looksLikeMarkdown(plain)) {
    return ensureNoteTableClass(markdownToNoteHtml(plain));
  }
  return ensureNoteTableClass(normalizeNoteHtml(content));
}

bool looksLikeMarkdown(String text) {
  final t = text.trim();
  if (t.isEmpty) return false;
  return RegExp(r'^#{1,3}\s', multiLine: true).hasMatch(t) ||
      RegExp(r'^\|.+\|', multiLine: true).hasMatch(t) ||
      RegExp(r'^[-*]\s+', multiLine: true).hasMatch(t) ||
      RegExp(r'^\d+\.\s+', multiLine: true).hasMatch(t) ||
      t.contains('**') ||
      t.contains('__');
}

String parseInlineMarkdown(String text) {
  var s = escapeHtml(text);
  s = s.replaceAllMapped(RegExp(r'`([^`]+)`'), (m) => '<code>${m[1]}</code>');
  s = s.replaceAllMapped(RegExp(r'\*\*(.+?)\*\*'), (m) => '<strong>${m[1]}</strong>');
  s = s.replaceAllMapped(RegExp(r'__(.+?)__'), (m) => '<strong>${m[1]}</strong>');
  s = s.replaceAllMapped(RegExp(r'\*(.+?)\*'), (m) => '<em>${m[1]}</em>');
  s = s.replaceAllMapped(
    RegExp(r'\[([^\]]+)\]\(([^)]+)\)'),
    (m) =>
        '<a href="${m[2]}" target="_blank" rel="noopener noreferrer">${m[1]}</a>',
  );
  return s;
}

bool _isTableRow(String line) {
  final t = line.trim();
  if (!t.contains('|')) return false;
  return t.split('|').where((c) => c.trim().isNotEmpty).length >= 2;
}

bool _isTableSeparator(String line) {
  return RegExp(r'^\|?[\s|:-]+\|?$').hasMatch(line.trim()) && line.contains('-');
}

bool _isHeadingLine(String line) => RegExp(r'^#{1,3}\s+').hasMatch(line.trim());

bool _isBulletLine(String line) => RegExp(r'^[-*]\s+').hasMatch(line.trim());

bool _isNumberedLine(String line) => RegExp(r'^\d+\.\s+').hasMatch(line.trim());

bool _isSectionLine(String line) =>
    RegExp(r'^[أ-يa-zA-Z]\)\s+').hasMatch(line.trim());

List<String> _parseTableRow(String line) {
  return line
      .trim()
      .replaceFirst(RegExp(r'^\|'), '')
      .replaceFirst(RegExp(r'\|$'), '')
      .split('|')
      .map((c) => c.trim())
      .toList();
}

String _renderTable(List<String> lines) {
  final rows =
      lines.where((l) => l.trim().isNotEmpty && !_isTableSeparator(l)).toList();
  if (rows.isEmpty) return '';
  final parsed = rows.map(_parseTableRow).toList();
  final header = parsed.first;
  final body = parsed.length > 1 ? parsed.sublist(1) : <List<String>>[];
  final thead =
      '<thead><tr>${header.map((c) => '<th>${parseInlineMarkdown(c)}</th>').join()}</tr></thead>';
  final tbody = body.isEmpty
      ? ''
      : '<tbody>${body.map((row) => '<tr>${row.map((c) => '<td>${parseInlineMarkdown(c)}</td>').join()}</tr>').join()}</tbody>';
  return '<table class="note-table">$thead$tbody</table>';
}

bool _isTableBlock(List<String> lines) {
  final meaningful = lines.where((l) => l.trim().isNotEmpty).toList();
  if (meaningful.length < 2) return false;
  return meaningful.every(_isTableRow) ||
      (meaningful.where(_isTableRow).length >= 2 &&
          meaningful.any(_isTableSeparator));
}

String _renderHeading(String line) {
  final trimmed = line.trim();
  final level = RegExp(r'^#+').firstMatch(trimmed)?.group(0)?.length ?? 1;
  final tag = 'h${level.clamp(1, 3)}';
  final text = trimmed.replaceFirst(RegExp(r'^#{1,3}\s+'), '');
  return '<$tag>${parseInlineMarkdown(text)}</$tag>';
}

String _renderBulletList(List<String> lines) {
  final items = lines
      .map((l) =>
          '<li>${parseInlineMarkdown(l.trim().replaceFirst(RegExp(r'^[-*]\s+'), ''))}</li>')
      .join();
  return '<ul>$items</ul>';
}

String _renderNumberedList(List<String> lines) {
  final items = lines
      .map((l) =>
          '<li>${parseInlineMarkdown(l.trim().replaceFirst(RegExp(r'^\d+\.\s+'), ''))}</li>')
      .join();
  return '<ol>$items</ol>';
}

String _renderParagraph(List<String> lines) {
  return '<p>${parseInlineMarkdown(lines.join('\n'))}</p>';
}

String markdownToNoteHtml(String markdown) {
  final text = markdown.replaceAll('\r\n', '\n').trim();
  if (text.isEmpty) return '<p></p>';

  final lines = text.split('\n');
  final html = <String>[];
  var i = 0;

  while (i < lines.length) {
    while (i < lines.length && lines[i].trim().isEmpty) {
      i++;
    }
    if (i >= lines.length) break;

    final line = lines[i];

    if (_isTableRow(line)) {
      final tableLines = <String>[];
      while (i < lines.length) {
        final current = lines[i];
        if (current.trim().isEmpty) {
          if (tableLines.isNotEmpty &&
              i + 1 < lines.length &&
              _isTableRow(lines[i + 1])) {
            i++;
            continue;
          }
          break;
        }
        if (_isTableRow(current) || _isTableSeparator(current)) {
          tableLines.add(current);
          i++;
          continue;
        }
        break;
      }
      if (_isTableBlock(tableLines)) {
        html.add(_renderTable(tableLines));
      } else if (tableLines.isNotEmpty) {
        html.add(_renderParagraph(tableLines));
      }
      continue;
    }

    if (_isHeadingLine(line)) {
      html.add(_renderHeading(line));
      i++;
      continue;
    }

    if (_isSectionLine(line)) {
      html.add('<h3>${parseInlineMarkdown(line.trim())}</h3>');
      i++;
      continue;
    }

    if (_isBulletLine(line)) {
      final listLines = <String>[];
      while (i < lines.length && _isBulletLine(lines[i])) {
        listLines.add(lines[i]);
        i++;
      }
      html.add(_renderBulletList(listLines));
      continue;
    }

    if (_isNumberedLine(line)) {
      final listLines = <String>[];
      while (i < lines.length && _isNumberedLine(lines[i])) {
        listLines.add(lines[i]);
        i++;
      }
      html.add(_renderNumberedList(listLines));
      continue;
    }

    final paraLines = <String>[];
    while (i < lines.length) {
      final current = lines[i];
      if (current.trim().isEmpty) break;
      if (_isTableRow(current) ||
          _isHeadingLine(current) ||
          _isBulletLine(current) ||
          _isNumberedLine(current) ||
          _isSectionLine(current)) {
        break;
      }
      paraLines.add(current);
      i++;
    }
    if (paraLines.isNotEmpty) {
      html.add(_renderParagraph(paraLines));
    }
  }

  final joined = html.where((b) => b.isNotEmpty).join('');
  return joined.isEmpty ? '<p></p>' : joined;
}

String extractTitleFromImportedContent(String content) {
  final lines = content.split('\n');
  for (final line in lines) {
    final t = line.trim();
    if (t.startsWith('# ')) return t.substring(2).trim();
    if (t.startsWith('## ')) return t.substring(3).trim();
  }
  final plain = content.trim().split('\n').first.trim();
  if (plain.length > 60) return '${plain.substring(0, 60)}…';
  return plain.isEmpty ? 'ملاحظة مستوردة' : plain;
}
